### PR TITLE
fix(detector): require a fence to close on its own marker (GH-236)

### LINKED ISSUE #236: Stop a ~~~ line inside a backtick fence from ending code protection in the validator
Item 2 from #143, never fixed: PR #145 touched only the OpenAI plugin validator. `FENCED_CODE` in `detector/validate.js` line 36 accepts either fence marker as the closer, regardless of which one opened the block:

```js
const FENCED_CODE = /^(?:```|~~~)[^\n]*\n[\s\S]*?^(?:```|~~~)[ \t]*$/gm;
```

```
$ node -e "
const {validate}=require('./detector/validate.js'); const o={skipResidual:true};
const orig='Text before.\n\n\`\`\`md\nexample line one\n~~~\nSECRET CODE A\n\`\`\`\n\nText after.';
const rew ='Text before.\n\n\`\`\`md\nexample line one\n~~~\nSECRET CODE B\n\`\`\`\n\nText after.';
console.log(JSON.stringify(validate(orig,rew,o).errors.map(e=>e.code)));"
[]
```

Code inside a fenced block was rewritten and no `code-block-modified` error fired. Writing a `~~~` sample inside a backtick fence is the normal way to document Markdown fences, so this fires on exactly the documents this repo produces. `fenceRanges()` in `detector/patterns.js` gets it right by tracking the opening fence character, so the two files disagree. The mirror case (a `~~~` block containing a backtick fence) has the same bug, and `scripts/self-scan.js` line 67 carries the same regex.

### Scope

`detector/validate.js` and the copy in `scripts/self-scan.js`, plus fixtures in `detector/validate.test.js`. A correct fix probably replaces the regex with a small line scanner that remembers the opening fence character and length (CommonMark requires the closer to be at least as long). Whether to also honor the #77 rule about info strings on closing fences is worth deciding first, which is why this is not a first issue. Run `bash scripts/sync-plugin-skill.sh` afterwards.

### Acceptance criteria

- A change inside a backtick-fenced block that contains a `~~~` line raises `code-block-modified`.
- The mirror case raises it too.
- Existing fenced-code fixtures still pass.
- `npm test` passes and `git diff` is empty after the sync script.

Add an `## [Unreleased]` changelog entry.

### PR DESCRIPTION
**Pull request by anujsingh-cse**
Closes #236

## Problem

`FENCED_CODE` in `detector/validate.js` accepted either fence marker as the closer, regardless of which one opened the block:

```js
const FENCED_CODE = /^(?:```|~~~)[^\n]*\n[\s\S]*?^(?:```|~~~)[ \t]*$/gm;
```

A `~~~` sample line inside a backtick fence (the normal way to document Markdown fences) terminated the block early, so code that followed was treated as prose and a rewrite of it produced no `code-block-modified` error. This is part 2 of #143 and was previously only half-fixed (PR #145 touched the OpenAI plugin validator, not `detector/validate.js`).

## Fix

Match the fence type explicitly so a fence only closes on its own marker:

```js
const FENCED_CODE =
  /^```[^\n]*\n[\s\S]*?^```[ \t]*$|^~~~[^\n]*\n[\s\S]*?^~~~[ \t]*$/gm;
```

No capture group is introduced, so `extractAll`'s `m[1]` handling is unaffected.

## Tests

Added two regression cases to `detector/validate.test.js`:

- a `~~~` line inside a backtick fence no longer ends it → editing the enclosed code fires `code-block-modified`
- a ``` line inside a tilde fence no longer ends it → same

`npm test` passes (validator, detector fixtures, and CATEGORIES.md contracts all green).

### CODE REVIEW

**Review by conorbronsdon — CHANGES_REQUESTED** [review]
Codex review, with a GPT-5.6 Sol reviewer, at Conor's request.

Thanks, @anujsingh-cse, for adding regressions in both mixed-marker directions and synchronizing the generated validators. Those cover the immediate example well. I still reproduced two acceptance gaps from #236 at this head:

- A four-backtick outer fence closes on an inner three-backtick line, so a later code edit incorrectly returns `ok: true`.
- scripts/self-scan.js retains the marker-agnostic matcher, leaving code after an opposite-marker example visible to the scan.

The required Unreleased changelog entry is also missing. Please use the line-scanner approach described in the issue so both paths track the opening marker and full run length, cover longer fences and self-scan, then regenerate the bundled validators. The current tests and CI pass, but don't cover these cases yet. This needs another implementation pass before merging.

**Comment by anujsingh-cse** [comment]
Thanks for the detailed review. I'll reshape the extraction into a small line scanner that tracks the opening fence marker *and* its length (matching the rule in CommonMark that the closer must be at least as long), so the 4-backtick-in-3(-backtick-in) nesting case closes correctly. I'll also apply the same fix to \scripts/self-scan.js\, add the missing \## [Unreleased]\ changelog entry, cover the scanner + the nested-fence case in tests, and regenerate the sync'd copies.

**Comment by anujsingh-cse** [comment]
All three review points are now addressed on the updated head (\333246\), and CI is green:

- **Four-backtick outer fence no longer closes on an inner three-backtick line.** Replaced the regex with a line scanner that tracks the opening marker and run length; per CommonMark a fence closes only on the same marker at equal or greater length (mirrors \enceRanges()\ in \detector/patterns.js\).
- **\scripts/self-scan.js\ uses the same scanner**, so content after an opposite-marker example is blanked correctly.
- **Added the \## [Unreleased]\ changelog entry** under \### Fixed\.

Added a regression test for the nested four-backtick case, re-synced the bundled validators, and kept both upstream changelog entries during the rebase. \
pm test\, \self-scan --check\, and CI (\check\/\ssot\/\	est\) all pass.

Could you take another look when you get a chance? Happy to adjust further.

**Comment by conorbronsdon** [comment]
thanks on it

File: detector/validate.js [hunk 1]
```diff
@@ -33,7 +33,10 @@
 
 const AIDetectorValidate = (() => {
   // ═══ Block extractors ══════════════════════════════════════════════
-  const FENCED_CODE = /^(?:```|~~~)[^\n]*\n[\s\S]*?^(?:```|~~~)[ \t]*$/gm;
+  // A fence closes only on its own marker: a `~~~` line inside a ``` block must
+// not end it (see the tilde-fence/backtick-fence cases in validate.test.js).
+const FENCED_CODE =
+  /^```[^\n]*\n[\s\S]*?^```[ \t]*$|^~~~[^\n]*\n[\s\S]*?^~~~[ \t]*$/gm;
```
[the remark this answers is not in the document]
Comment (conorbronsdon) [reply detector/validate.js]: Resolved in e38a357: self-scan now uses the same marker/run-length scanner and has mixed-marker, longer-outer-fence, indentation, and CRLF coverage.

[the remark this answers is not in the document]
Comment (conorbronsdon) [reply detector/validate.js]: Resolved at the current head: CHANGELOG.md includes the Unreleased fence-preservation fix for #236.

[the remark this answers is not in the document]
Comment (conorbronsdon) [reply detector/validate.js]: Resolved in e38a357: the scanner records the opener marker and full run length, and only a same-marker run of at least that length can close it. The four-backtick/three-backtick regression is covered.

[the remark this answers is not in the document]
Comment (conorbronsdon) [reply detector/validate.js]: Resolved in e38a357: the canonical validator and all three generated copies were synchronized and verified byte-identical; the full package parity/test suite passes.

**Review by conorbronsdon — APPROVED** [review]
Approved at e38a357 after four repair/review cycles. Verified the reported indentation and CRLF edge cases, positional self-scan masking, same-marker/full-run closing semantics, generated-copy parity, 20/20 local test groups, self-scan, green CI, authenticated Claude review, isolated free Muse review, and zero open code-scanning alerts. Remaining free-review notes concern pre-existing detector-wide CommonMark limitations outside #236, not regressions in this change.
